text/0000-unit-from-iterator.md

```diff
@@ -0,0 +1,84 @@
+- Feature Name: unit_from_iterator
+- Start Date: 2015-05-20
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Implement `std::iter::FromIterator<()>` for the empty tuple type `()`, also known as “unit”.
+
+# Motivation
+
+The `Result` type implements `FromIterator` in a way that creates a result out of an iterable of results, extracting the first error if any, and otherwise returning `Ok` with the contents of all values:
+
+```rust
+println!("{:?}", Result::<Vec<i32>, MyError>::from_iter(vec![Ok(4), Ok(7), Ok(-3903)])); // Ok([4, 7, -3903])
+println!("{:?}", Result::<Vec<i32>, MyError>::from_iter(vec![Ok(5), Err(MyError::DivisionByZero), Ok(842), Err(MyError::Overflow)])); // Err(MyError::DivisionByZero)
+```
+
+Implementing this RFC would allow this pattern to be used with iterables whose item type is of the form `Result<(), T>`.
+
+For example, suppose we have a function which moves all values from a `mpsc::Receiver` into a `mpsc::Sender`. Currently, this could be written as follows:
+
+```rust
+fn forward_values<T>(src: Receiver<T>, dst: Sender<T>) -> Result<(), SendError<T>> {
+    src.iter().map(|val| dst.send(val)).fold(Ok(()), Result::and))
+}
+```
+
+This has the flaw of exhausting the receiver even after an error is encountered. With the proposed trait implementation, it could be refactored into the following:
+
+```rust
+fn forward_values<T>(src: Receiver<T>, dst: Sender<T>) -> Result<(), SendError<T>> {
+    src.iter().map(|val| dst.send(val)).collect()
+}
+```
+
+This version of the function immediately returns when the first error is encountered.
+
+# Detailed design
+
+Implement the trait `std::iter::FromIterator<()>` for the primitive type `()`.
+
+The implementation is very short:
```

Comment:
Good question. If the exhausting implementation is chosen, the generic version seems strictly better as far as I can tell. It could be used, for example, to collect any `Result<T, E>` into a `Result<(), E>`.